### ISSUE TITLE
Missing member ID split.

### ISSUE THREAD

**Issue by brian-reichle** [DotNetAnalyzers/ReflectionAnalyzers#135]
If we know exactly what type we will be reflecting against at analysis time, then it should be reported as REFL0003

eg.

```csharp
void Foo(SealedType x) => x.GetType().GetMethod("NotInSealed"); // we know it won't be a subclass because it's sealed.
void Bar(AbstractType y) => typeof(AbstractType).GetMethod("NotInAbstract"); // we know it won't be a subclass because we explicitly provided an exact type.
```

If we only know a base class and not the actual type that will be used at runtime, then report missing members as REFL0009

eg.

```csharp
void Foo(AbstractType x) => x.GetType().GetMethod("NotInAbstract"); // the actual type will be a subclas of "AbstractType", but we can't be certain which.
```

**Comment by brian-reichle** [comment]
Delegates are a bit of a special case, we know that all *concrete* delegates will define an `Invoke` method, because it's required by ECMA-335 partition II section 14.6. (MulticastDelegate seems to be the only non-sealed subclass of Delegate.)

**Comment by jnm2** [comment]
Other conventionally generated delegate methods to look for are BeginInvoke, EndInvoke, and a constructor.

**Comment by JohanLarsson** [comment]
More cases when we are sure:
```cs
this.GetType().GetMethod("Missing") 
x.GetType().GetMethod("Missing") // when x is created with `new` in current block.
```

**Comment by brian-reichle** [comment]
Technically, the clr spec lists BeginInvoke & EndInvoke as optional ... but probably a safe bet anyway. The delegate is required to provide a constructor that takes an object and an IntPtr, and these 4 members are the only members permitted by the spec.

**Comment by jnm2** [comment]
Begin/EndInvoke throw PNSE on .NET Core anyway. :D

**Comment by brian-reichle** [comment]
`this.GetType()` is no more reliable than calling `GetType()` on a parameter. If the current type is not sealed then "this" could just as easily be a subclass.

If you see the type being constructed, then you (mostly*) know what calling `GetType()` on it will return, but in this case I would normally prefer to be explicit and use typeof.

(* calling `new X().GetType()` will be less reliable if `X` inherits from `ContextBoundObject` as the object creation may be intercepted and a [different type returned](https://dotnetfiddle.net/qWDKxM) - this probably doesn't apply to .NET Core though)

**Comment by JohanLarsson** [comment]
Ok, closing this now then, we special case `typeof(...)` now. We can always tweak things later.
